app/Listeners/CheckoutableListener.php

```diff
@@ -0,0 +1,184 @@
+use App\Notifications\CheckoutLicenseNotification;
+use App\Notifications\CheckoutLicenseSeatNotification;
+use Illuminate\Support\Facades\Notification;
+
+class CheckoutableListener
+{
+
+    /**
+     * Notify the user about the checked out checkoutable
+     */
+    public function onCheckedOut($event) {
+        /**
+         * When the item wasn't checked out to a user, we can't send notifications
+         */
+        if(! $event->checkedOutTo instanceof User) {
+            return;
+        }
+
+        /**
+         * Make a checkout acceptance and attach it in the notification
+         */
+        $acceptance = $this->getCheckoutAcceptance($event);       
+
+        Notification::send(
+            $this->getNotifiables($event), 
+            $this->getCheckoutNotification($event, $acceptance)
+        );
+    }
+
+    /**
+     * Notify the user about the checked in checkoutable
+     */    
+    public function onCheckedIn($event) {
+        /**
+         * When the item wasn't checked out to a user, we can't send notifications
+         */
+        if(!$event->checkedOutTo instanceof User) {
+            return;
+        }
+
+        /**
+         * Send the appropriate notification
+         */
+        Notification::send(
+            $this->getNotifiables($event), 
+            $this->getCheckinNotification($event)
+        );
+    }      
+
+    /**
+     * Generates a checkout acceptance
+     * @param  Event $event
+     * @return mixed
+     */
+    private function getCheckoutAcceptance($event) {
+        if (!$event->checkoutable->requireAcceptance()) {
+            return null;
+        }
+
+        $acceptance = new CheckoutAcceptance;
+        $acceptance->checkoutable()->associate($event->checkoutable);
+        $acceptance->assignedTo()->associate($event->checkedOutTo);
+        $acceptance->save();
+
+        return $acceptance;      
+    }
+
+    /**
+     * Gets the entities to be notified of the passed event
+     * 
+     * @param  Event $event
+     * @return Collection
+     */
+    private function getNotifiables($event) {
+        $notifiables = collect();
+
+        /**
+         * Notify the user who checked out the item
+         */
+        $notifiables->push($event->checkedOutTo);
+
+        /**
+         * Notify Admin users if the settings is activated
+         */
+        if (Setting::getSettings()->admin_cc_email != '') {
+            $notifiables->push(new AdminRecipient());
+        }
+
+        return $notifiables;       
+    }
+
+    /**
+     * Get the appropriate notification for the event
+     * 
+     * @param  CheckoutableCheckedIn $event 
+     * @return Notification
+     */
+    private function getCheckinNotification($event) {
+
+        $model = get_class($event->checkoutable);
+
+        $notificationClass = null;
+
+        switch (get_class($event->checkoutable)) {
+            case Accessory::class:
+                $notificationClass = CheckinAccessoryNotification::class;
```

Comment:
Tiny nit - I tend to feel 'weird' when I see `case` statements looking at classnames. Instead, we should try and use the object system to avoid them, when possible.

In this case, could we have a class property on the particular Checkoutable class? e.g.

```php
class Accessory {
  // ...
  public static $checkinNotificationClass = CheckinAccessoryNotification::class;
  // ...
}
```

(etc)?

Reply:
Oh, I didn't explain the rest of it - the idea would be that then, this collapses into one line:

```php
return new self::$checkinNotificationClass($event->checkoutable, $event->checkedOutTo, // etc...
```